### PR TITLE
The Big Refactor

### PR DESCRIPTION
**Pull request by Amaras**
This PR aims to allow an easier understanding and ability to change the code in the future.
The changes are currently not ready to merge, and not tested yet.

This is a summary of the current state of changes (at the time of writing):
- [x] Used 3.10 syntax to remove Optional[T] and Union[T, U] in code
- [x] Refactor to minify the typing dependency in all the classes in `data/`
- [x] Changed some classes into dataclasses
- [x] Solved some mypy errors

The plan for the future is based on this talk by Conor Hoekstra: https://www.youtube.com/watch?v=nXZQfdxWgh0
In summary for our case:
- [x] Use Enum (especially for sets of related bare constants)
- [x] ~~Maybe use Optional (from optional.py)~~ decided against it because of no `__class_getitem__`
- [ ] Minimize the "Initialize Then Modify" pattern
- [x] (Not related to the talk) Use direct assignment instead of the `set_` methods when possible, and add warnings for the previous API.

There might be more refactors if applicable (for example: extract out the data classes creation to the classes as classmethods).

### CODE REVIEW

**Review by pascalaubry — APPROVED** [review]

**Review by pascalaubry — COMMENTED** [review]
Why did Player.__board_number disappear?

**Review by pascalaubry — APPROVED** [review]

**Review by pascalaubry — APPROVED** [review]
